File: backend/exams/sat_rules.py
"""
SAT Simulation Rule Engine — single authoritative source of truth.

This module defines ALL official Digital SAT structural constraints used by:
  - Backend publish validation (publish_service.py)
  - Backend question serializer validation
  - Frontend mirrors (frontend/src/lib/satRules.ts — keep in sync)

NEVER scatter SAT-specific constants across models, views, or serializers.
Import from here.

Official SAT structure (Digital SAT, 2024+):
  Reading & Writing: 2 modules × 27 questions × 32 minutes each
  Math:              2 modules × 22 questions × 35 minutes each

See: https://satsuite.collegeboard.org/digital/whats-on-the-test/structure
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .models import MockExam, Module, PracticeTest

# ── Official SAT subjects ────────────────────────────────────────────────────

SAT_SUBJECTS = ("READING_WRITING", "MATH")

# ── Per-module question counts (official Digital SAT) ───────────────────────

SAT_MODULE_QUESTION_COUNT: dict[str, int] = {
    "READING_WRITING": 27,
    "MATH": 22,
}

# ── Per-module time limits in minutes (official Digital SAT) ────────────────

SAT_MODULE_TIME_LIMIT_MINUTES: dict[str, int] = {
    "READING_WRITING": 32,
    "MATH": 35,
}

# ── Section order (deprecated — students can now start any section) ─────────
# Kept for reference only; enforcement removed from views.py.

SAT_SECTION_ORDER: list[str] = ["READING_WRITING", "MATH"]

# ── Break between sections (deprecated — break enforcement removed) ─────────
# Was: 600 seconds (10 minutes). No longer enforced.

SAT_BREAK_SECONDS: int = 0

# ── Allowed question_type values per section subject ────────────────────────
#
# Reading & Writing questions must carry type READING or WRITING.
# Math questions must carry type MATH.
# Mixing subjects in a module is an authoring error.

ALLOWED_QUESTION_TYPES_PER_SUBJECT: dict[str, frozenset[str]] = {
    "READING_WRITING": frozenset({"READING", "WRITING"}),
    "MATH": frozenset({"MATH"}),
}

# ── Module count per section (always exactly 2 for full SAT) ────────────────

SAT_MODULES_PER_SECTION = 2

# ── Full mock section requirements ─────────────────────────────────────────
#
# A full MOCK_SAT must contain BOTH subjects to faithfully simulate
# the official SAT experience (R&W section + Math section).

SAT_FULL_MOCK_REQUIRED_SUBJECTS = frozenset({"READING_WRITING", "MATH"})

# ── Score architecture (SAT 200–800 per section) ────────────────────────────
#
# Official Digital SAT scoring:
#
#   Reading & Writing section:
#     Base score:    200
#     Module 1 cap:  330  (a perfect M1 gives 200+330 = 530)
#     Module 2 cap:  270  (both perfect gives 200+330+270 = 800)
#
#   Math section:
#     Base score:    200
#     Module 1 cap:  380  (a perfect M1 gives 200+380 = 580)
#     Module 2 cap:  220  (both perfect gives 200+380+220 = 800)
#
# SCORING IS PROPORTIONAL:
#   contribution = round(correct_pts / total_pts × module_cap)
# This guarantees a perfect module always reaches its cap regardless of
# whether individual question scores are 10, 20, or 40 points.

SAT_SECTION_BASE_SCORE = 200
SAT_SECTION_MAX_SCORE = 800

# subject → {module_order → max contribution above base}
SAT_MODULE_SCORE_CAP: dict[str, dict[int, int]] = {
    "READING_WRITING": {1: 330, 2: 270},
    "MATH": {1: 380, 2: 220},
}

# subject → maximum section score when only that module is complete
SAT_MODULE_MAX_SECTION_SCORE: dict[str, dict[int, int]] = {
    # M1 perfect alone: 200 + M1_cap (M2 = 0)
    "READING_WRITING": {1: 530, 2: 800},
    "MATH":            {1: 580, 2: 800},
}

# ── Violation dataclass ──────────────────────────────────────────────────────


@dataclass(frozen=True)
class SatViolation:
    """
    A single structural violation of the SAT simulation rules.

    ``code``    — machine-readable identifier (used in tests / metrics).
    ``message`` — human-readable admin-facing description.
    ``blocking``— True when this prevents publish; False for warnings only.
    """

    code: str
    message: str
    blocking: bool = True


# ── Module-level validators ──────────────────────────────────────────────────


def validate_module(module: "Module", subject: str) -> list[SatViolation]:
    """
    Return all SAT violations for a single module row.

    ``subject`` is the parent PracticeTest.subject (READING_WRITING | MATH).
    """
    violations: list[SatViolation] = []

    questions = list(module.questions.all())
    q_count = len(questions)
    required_count = SAT_MODULE_QUESTION_COUNT.get(subject)

    if required_count is not None and q_count != required_count:
        subj_label = "Reading & Writing" if subject == "READING_WRITING" else "Math"
        violations.append(
            SatViolation(
                code="MODULE_QUESTION_COUNT",
                message=(
                    f"{subj_label} Module {module.module_order} requires exactly "
                    f"{required_count} questions (currently {q_count})."
                ),
                blocking=True,
            )
        )

    allowed_types = ALLOWED_QUESTION_TYPES_PER_SUBJECT.get(subject)
    if allowed_types:
        bad_type_qs = [q for q in questions if q.question_type not in allowed_types]
        if bad_type_qs:
            bad_types = sorted({q.question_type for q in bad_type_qs})
            allowed_display = "/".join(sorted(allowed_types))
            subj_label = "Reading & Writing" if subject == "READING_WRITING" else "Math"
            violations.append(
                SatViolation(
                    code="MODULE_QUESTION_TYPE",
                    message=(
                        f"{subj_label} Module {module.module_order} contains "
                        f"{len(bad_type_qs)} question(s) with invalid type "
                        f"({', '.join(bad_types)}). "
                        f"Only {allowed_display} allowed."
                    ),
                    blocking=True,
                )
            )

    return violations


# ── Section-level validators ─────────────────────────────────────────────────


def validate_practice_test(pt: "PracticeTest") -> list[SatViolation]:
    """
    Return all SAT violations for a PracticeTest (one section: R&W or Math).
    """
    violations: list[SatViolation] = []
    subject = getattr(pt, "subject", None)

    if subject not in SAT_SUBJECTS:
        violations.append(
            SatViolation(
                code="SECTION_INVALID_SUBJECT",
                message=f"Section subject '{subject}' is not a valid SAT subject.",
                blocking=True,
            )
        )
        return violations

    modules = list(pt.modules.all().order_by("module_order"))

    if len(modules) != SAT_MODULES_PER_SECTION:
        subj_label = "Reading & Writing" if subject == "READING_WRITING" else "Math"
        violations.append(
            SatViolation(
                code="SECTION_MODULE_COUNT",
                message=(
                    f"{subj_label} section must have exactly {SAT_MODULES_PER_SECTION} modules "
                    f"(currently {len(modules)})."
                ),
                blocking=True,
            )
        )
    else:
        for m in modules:
            violations.extend(validate_module(m, subject))

    return violations


# ── Mock-exam-level validators ───────────────────────────────────────────────


def validate_mock_exam(exam: "MockExam") -> list[SatViolation]:
    """
    Return all SAT violations for a MockExam (timed mock or midterm).

    Full MOCK_SAT: strict structural validation (both sections, correct counts).
    Midterm: relaxed — only validates that modules have questions; no fixed count.
    """
    from .models import MockExam as MockExamModel

    violations: list[SatViolation] = []
    tests = list(exam.tests.all())

    if exam.kind == MockExamModel.KIND_MIDTERM:
        # Midterms are institution-controlled: question counts are flexible.
        if len(tests) != 1:
            violations.append(
                SatViolation(
                    code="MIDTERM_SECTION_COUNT",
                    message="Midterm must have exactly one section.",
                    blocking=True,
                )
            )
            return violations
        pt = tests[0]
        need_mods = max(1, min(2, exam.midterm_module_count or 1))
        mods = list(pt.modules.all().order_by("module_order"))
        if len(mods) < need_mods:
            violations.append(
                SatViolation(
                    code="MIDTERM_MISSING_MODULES",
                    message=f"Midterm needs {need_mods} module(s) with questions.",
                    blocking=True,
                )
            )
        else:
            for m in mods[:need_mods]:
                if not list(m.questions.all()):
                    violations.append(
                        SatViolation(
                            code="MIDTERM_EMPTY_MODULE",
                            message=f"Module {m.module_order} must have at least one question.",
                            blocking=True,
                        )
                    )
        return violations

    # MOCK_SAT — full Digital SAT simulation: strict structural validation.
    if len(tests) == 0:
        violations.append(
            SatViolation(
                code="MOCK_NO_SECTIONS",
                message="Add both a Reading & Writing section and a Math section.",
                blocking=True,
            )
        )
        return violations

    subjects_present = {t.subject for t in tests}
    for required_subj in SAT_FULL_MOCK_REQUIRED_SUBJECTS:
        if required_subj not in subjects_present:
            label = "Reading & Writing" if required_subj == "READING_WRITING" else "Math"
            violations.append(
                SatViolation(
                    code="MOCK_MISSING_SECTION",
                    message=(
                        f"Full SAT mock requires a {label} section. "
                        f"Both sections must be present for simulation integrity."
                    ),
                    blocking=True,
                )
            )

    for pt in tests:
        violations.extend(validate_practice_test(pt))

    return violations


# ── Convenience: publish-ready check ────────────────────────────────────────


def mock_exam_publish_violations(exam: "MockExam") -> list[SatViolation]:
    """All blocking violations that prevent publishing this mock exam."""
    return [v for v in validate_mock_exam(exam) if v.blocking]


def practice_test_publish_violations(section: "PracticeTest") -> list[SatViolation]:
    """All blocking violations that prevent publishing this standalone section."""
    return [v for v in validate_practice_test(section) if v.blocking]


def compute_sat_module_score(
    earned_points: int | float,
    total_possible_points: int | float,
    subject: str,
    module_order: int,
) -> int:
    """
    Proportional SAT module score contribution.

    Maps weighted question correctness to the module's official point contribution:
      - Math M1 cap   = 380 → full section floor at 200+380 = 580 when perfect
      - Math M2 cap   = 220 → full section ceiling at 800 when both perfect
      - R&W M1 cap    = 330 → floor 530 when perfect
      - R&W M2 cap    = 270 → ceiling 800 when both perfect

    Using proportional scoring means the `score` field on Question acts as a
    relative weight (harder questions can count more), while a perfect module
    always yields the full cap regardless of whether questions are worth 10, 20,
    or 40 points each.

    Returns 0 when total_possible_points == 0 (empty module).
    """
    caps = SAT_MODULE_SCORE_CAP.get(subject, {})
    cap = caps.get(module_order, 0)
    if not cap or not total_possible_points:
        return 0
    fraction = max(0.0, min(1.0, float(earned_points) / float(total_possible_points)))
    return min(int(round(fraction * cap)), cap)


def is_question_type_allowed(question_type: str, subject: str) -> bool:
    """True if question_type is valid for the given section subject."""
    allowed = ALLOWED_QUESTION_TYPES_PER_SUBJECT.get(subject)
    if allowed is None:
        return True  # Unknown subject — don't block
    return question_type in allowed


def allowed_question_types_for_subject(subject: str) -> list[str]:
    """Ordered list of allowed question_type values for the given subject."""
    allowed = ALLOWED_QUESTION_TYPES_PER_SUBJECT.get(subject)
    if allowed is None:
        return ["MATH", "READING", "WRITING"]
    # Return in a stable, human-friendly order
    order = ["MATH", "READING", "WRITING"]
    return [t for t in order if t in allowed]

File: frontend/src/lib/satRules.ts
/**
 * SAT Simulation Rule Engine — canonical frontend source of truth.
 *
 * This module mirrors backend/exams/sat_rules.py.
 * When changing SAT structural rules, update BOTH files.
 *
 * Official Digital SAT structure (2024+):
 *   Reading & Writing: 2 modules × 27 questions × 32 minutes each
 *   Math:              2 modules × 22 questions × 35 minutes each
 */

// ── Official SAT subjects ────────────────────────────────────────────────────

export type SatSubject = "READING_WRITING" | "MATH";

export const SAT_SUBJECTS: readonly SatSubject[] = ["READING_WRITING", "MATH"];

export function isSatSubject(s: string): s is SatSubject {
  return s === "READING_WRITING" || s === "MATH";
}

// ── Question types ──────────────────────────────────────────────────────────

export type SatQuestionType = "MATH" | "READING" | "WRITING";

export const ALL_QUESTION_TYPES: readonly SatQuestionType[] = ["MATH", "READING", "WRITING"];

// ── Per-module question counts ───────────────────────────────────────────────

export const SAT_MODULE_QUESTION_COUNT: Record<SatSubject, number> = {
  READING_WRITING: 27,
  MATH: 22,
} as const;

// ── Per-module time limits (minutes) ────────────────────────────────────────

export const SAT_MODULE_TIME_LIMIT_MINUTES: Record<SatSubject, number> = {
  READING_WRITING: 32,
  MATH: 35,
} as const;

// ── Allowed question types per section subject ───────────────────────────────

export const ALLOWED_QUESTION_TYPES: Record<SatSubject, readonly SatQuestionType[]> = {
  READING_WRITING: ["READING", "WRITING"],
  MATH: ["MATH"],
} as const;

// ── Modules per section ──────────────────────────────────────────────────────

export const SAT_MODULES_PER_SECTION = 2;

// ── Score architecture ───────────────────────────────────────────────────────
//
// Official Digital SAT scoring (must match backend sat_rules.py exactly):
//
//   Reading & Writing:
//     Base: 200  |  M1 cap: 330 (floor 530)  |  M2 cap: 270 (ceiling 800)
//   Math:
//     Base: 200  |  M1 cap: 380 (floor 580)  |  M2 cap: 220 (ceiling 800)
//
// Scoring is PROPORTIONAL:
//   contribution = round(correct_pts / total_pts × module_cap)
// A perfect module always yields the full cap regardless of question score weights.

export const SAT_SECTION_BASE_SCORE = 200;
export const SAT_SECTION_MAX_SCORE = 800;

/** Maximum contribution per module above the 200 base. */
export const SAT_MODULE_SCORE_CAP: Record<SatSubject, Record<1 | 2, number>> = {
  READING_WRITING: { 1: 330, 2: 270 },
  MATH:            { 1: 380, 2: 220 },
} as const;

/**
 * Maximum section score when only through Module N with perfection.
 * E.g. Math M1 perfect alone = 580; both perfect = 800.
 */
export const SAT_MODULE_MAX_SECTION_SCORE: Record<SatSubject, Record<1 | 2, number>> = {
  READING_WRITING: { 1: 530, 2: 800 },
  MATH:            { 1: 580, 2: 800 },
} as const;

/**
 * Compute proportional SAT module score contribution.
 * Mirrors backend sat_rules.compute_sat_module_score().
 *
 * ``earnedPoints``   — weighted sum of scores for correct questions
 * ``totalPoints``    — weighted sum of scores for all questions
 * ``subject``        — READING_WRITING | MATH
 * ``moduleOrder``    — 1 or 2
 */
export function computeSatModuleScore(
  earnedPoints: number,
  totalPoints: number,
  subject: string | undefined | null,
  moduleOrder: 1 | 2,
): number {
  if (!subject || !isSatSubject(subject) || !totalPoints) return 0;
  const cap = SAT_MODULE_SCORE_CAP[subject][moduleOrder];
  if (!cap) return 0;
  const fraction = Math.max(0, Math.min(1, earnedPoints / totalPoints));
  return Math.min(Math.round(fraction * cap), cap);
}

/**
 * Compute the full SAT section score (200–800) from per-module earned points.
 * Mirrors the logic in TestAttempt.complete_test().
 *
 * ``modules`` — array of {earnedPoints, totalPoints, moduleOrder}
 */
export function computeSatSectionScore(
  subject: string | undefined | null,
  modules: Array<{ earnedPoints: number; totalPoints: number; moduleOrder: 1 | 2 }>,
): number {
  let total = SAT_SECTION_BASE_SCORE;
  for (const m of modules) {
    total += computeSatModuleScore(m.earnedPoints, m.totalPoints, subject, m.moduleOrder);
  }
  return Math.min(total, SAT_SECTION_MAX_SCORE);
}

// ── Subject display labels ───────────────────────────────────────────────────

export const SAT_SUBJECT_LABEL: Record<SatSubject, string> = {
  READING_WRITING: "Reading & Writing",
  MATH: "Mathematics",
} as const;

export const SAT_QUESTION_TYPE_LABEL: Record<SatQuestionType, string> = {
  MATH: "Math",
  READING: "Reading",
  WRITING: "Writing",
} as const;

// ── Helper utilities ─────────────────────────────────────────────────────────

/**
 * Returns the allowed question types for a given section subject.
 * Falls back to all types for unknown subjects.
 */
export function allowedQuestionTypesForSubject(
  subject: string | undefined | null,
): readonly SatQuestionType[] {
  if (!subject || !isSatSubject(subject)) return ALL_QUESTION_TYPES;
  return ALLOWED_QUESTION_TYPES[subject];
}

/**
 * Returns true if the question_type is valid for the given section subject.
 */
export function isQuestionTypeAllowed(
  questionType: string,
  subject: string | undefined | null,
): boolean {
  if (!subject || !isSatSubject(subject)) return true;
  return (ALLOWED_QUESTION_TYPES[subject] as readonly string[]).includes(questionType);
}

/**
 * Returns the required question count for a given subject.
 * Returns null for unknown subjects (no constraint enforced).
 */
export function requiredQuestionCount(subject: string | undefined | null): number | null {
  if (!subject || !isSatSubject(subject)) return null;
  return SAT_MODULE_QUESTION_COUNT[subject];
}

/**
 * Returns the required time limit (minutes) for a given subject.
 */
export function requiredTimeLimitMinutes(subject: string | undefined | null): number | null {
  if (!subject || !isSatSubject(subject)) return null;
  return SAT_MODULE_TIME_LIMIT_MINUTES[subject];
}

// ── Module composition violation types ──────────────────────────────────────

export type ModuleViolation = {
  code: string;
  message: string;
};

/**
 * Validate the composition of questions in a module against SAT rules.
 * Returns an array of violations (empty = valid).
 *
 * ``questions`` — minimal question shape: {question_type: string}
 * ``subject``   — the PracticeTest.subject for this module
 * ``moduleOrder`` — 1 or 2 (for human-readable messages)
 */
export function validateModuleComposition(
  questions: Array<{ question_type: string }>,
  subject: string | undefined | null,
  moduleOrder: number = 1,
): ModuleViolation[] {
  if (!subject || !isSatSubject(subject)) return [];

  const violations: ModuleViolation[] = [];
  const required = SAT_MODULE_QUESTION_COUNT[subject];
  const subjectLabel = SAT_SUBJECT_LABEL[subject];

  if (questions.length !== required) {
    violations.push({
      code: "MODULE_QUESTION_COUNT",
      message:
        `${subjectLabel} Module ${moduleOrder} requires exactly ${required} questions ` +
        `(currently ${questions.length}).`,
    });
  }

  const allowed = ALLOWED_QUESTION_TYPES[subject] as readonly string[];
  const wrongType = questions.filter((q) => !allowed.includes(q.question_type));
  if (wrongType.length > 0) {
    const badTypes = [...new Set(wrongType.map((q) => q.question_type))].join(", ");
    const allowedDisplay = allowed.join("/");
    violations.push({
      code: "MODULE_QUESTION_TYPE",
      message:
        `${subjectLabel} Module ${moduleOrder} contains ${wrongType.length} question(s) ` +
        `with invalid type (${badTypes}). Only ${allowedDisplay} allowed.`,
    });
  }

  return violations;
}

// ── Progress helpers for authoring UI ───────────────────────────────────────

export type ModuleProgress = {
  /** Current question count */
  current: number;
  /** Required question count for this subject (null = no constraint) */
  required: number | null;
  /** Fraction 0–1 toward requirement (null if no constraint) */
  fraction: number | null;
  /** True when current === required */
  complete: boolean;
  /** True when current > required */
  over: boolean;
  /** Display string e.g. "19 / 22" */
  label: string;
};

export function getModuleProgress(
  currentCount: number,
  subject: string | undefined | null,
): ModuleProgress {
  const required = requiredQuestionCount(subject);

  if (required === null) {
    return {
      current: currentCount,
      required: null,
      fraction: null,
      complete: false,
      over: false,
      label: String(currentCount),
    };
  }

  const complete = currentCount === required;
  const over = currentCount > required;
  const fraction = Math.min(currentCount / required, 1);

  return {
    current: currentCount,
    required,
    fraction,
    complete,
    over,
    label: `${currentCount} / ${required}`,
  };
}

// ── Type mismatch warnings ───────────────────────────────────────────────────

/**
 * Returns a warning message if the question_type is wrong for the subject,
 * or null if valid.
 */
export function questionTypeWarning(
  questionType: string,
  subject: string | undefined | null,
): string | null {
  if (!subject || !isSatSubject(subject)) return null;
  if (isQuestionTypeAllowed(questionType, subject)) return null;

  const allowed = allowedQuestionTypesForSubject(subject)
    .map((t) => SAT_QUESTION_TYPE_LABEL[t])
    .join(" or ");
  const subjectLabel = SAT_SUBJECT_LABEL[subject];
  const typeName = SAT_QUESTION_TYPE_LABEL[questionType as SatQuestionType] ?? questionType;

  return (
    `${typeName} questions are not allowed in a ${subjectLabel} module. ` +
    `Use ${allowed} question types here.`
  );
}
